p2p/ipld/read.go

```diff
@@ -3,16 +3,215 @@ package ipld
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 
 	"github.com/ipfs/go-cid"
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
 	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
 )
 
-// /////////////////////////////////////
+// //////////////////////////////////////
+//	Retrieve Block Data
+// ////////////////////////////////////
+
+const baseErrorMsg = "failure to retrieve block data:"
+
+var ErrEncounteredTooManyErrors = fmt.Errorf("%s %s", baseErrorMsg, "encountered too many errors")
+var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
+
+// RetrieveBlockData asynchronously fetches block data until the underlying extended
+// data square can be restored and returned.
+func RetrieveBlockData(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	api coreiface.CoreAPI,
+	codec rsmt2d.CodecType,
+) (types.Data, error) {
+	edsWidth := uint32(len(dah.ColumnRoots))
+	originalSquareWidth := edsWidth / 2
+
+	// keep track of leaves using thread safe counter
+	sc := newshareCounter(ctx, edsWidth)
+
+	// convert the row and col roots into Cids
+	rowRoots := dah.RowsRoots.Bytes()
+	rowCids, err := rootsToCids(rowRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	colRoots := dah.ColumnRoots.Bytes()
+	colCids, err := rootsToCids(colRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	// async attempt to fetch each leaf
+	for axisIdx := uint32(0); axisIdx < uint32(len(rowCids)); axisIdx++ {
+		for idx := uint32(0); idx < edsWidth; idx++ {
+			go sc.retrieveShare(colCids[axisIdx], false, axisIdx, idx, api)
+			go sc.retrieveShare(rowCids[axisIdx], true, axisIdx, idx, api)
```

Comment:
Thinking about it, there might be a more clever way in the ipfs API to retrieve a bunch of leaves (similar to the batch getting of leaves) in parallel. IMO, we could look into this but not necessarily in this PR.

Reply:
Yessss, I was just about to look at this

Reply:
I will create an issue with an alternative approach, as that can be treated as an optimization

Reply:
created this: https://github.com/lazyledger/lazyledger-core/issues/278